src/Components/Blazor/Validation/src/ObjectGraphDataAnnotationsValidator.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    public class ObjectGraphDataAnnotationsValidator : ComponentBase
+    {
+        private static readonly object ValidationContextValidatorKey = new object();
+        private static readonly object ValidatedObjectsKey = new object();
+        private ValidationMessageStore _validationMessageStore;
+
+        [CascadingParameter]
+        internal EditContext EditContext { get; set; }
+
+        protected override void OnInitialized()
+        {
+            _validationMessageStore = new ValidationMessageStore(EditContext);
+
+            // Perform object-level validation (starting from the root model) on request
+            EditContext.OnValidationRequested += (sender, eventArgs) =>
+            {
+                _validationMessageStore.Clear();
+                ValidateObject(EditContext.Model, new HashSet<object>());
+                EditContext.NotifyValidationStateChanged();
+            };
+
+            // Perform per-field validation on each field edit
+            EditContext.OnFieldChanged += (sender, eventArgs) =>
+                ValidateField(EditContext, _validationMessageStore, eventArgs.FieldIdentifier);
+        }
+
+        internal void ValidateObject(object value, HashSet<object> visited)
+        {
+            if (value is null)
+            {
+                return;
+            }
+
+            if (!visited.Add(value))
+            {
+                // Already visited this object.
+                return;
+            }
+
+            if (value is IEnumerable<object> enumerable)
+            {
+                var index = 0;
+                foreach (var item in enumerable)
+                {
+                    ValidateObject(item, visited);
+                    index++;
+                }
+
+                return;
+            }
+
+            var validationResults = new List<ValidationResult>();
```

Comment:
You need to add the following code....
```
if (value.GetType().Assembly == typeof(object).Assembly)
  return;
```

Otherwise you will validate every string in a `string[]`, and then every character of every string.